tests/41end-to-end-keys/07-backup.pl

```diff
@@ -0,0 +1,271 @@
+            first_message_index => 1,
+            forwarded_count     => 0,
+            is_verified         => JSON::false,
+            session_data        => "anotheropaquestring",
+         }
+      )->then( sub {
+         my ( $content ) = @_;
+         log_if_fail "Content", $content;
+
+         do_request_json_for( $user,
+            method  => "GET",
+            uri     => "/unstable/room_keys/keys/!abcd/1234",
+            params  => {
+               version => $current_version,
+            }
+         );
+      })->then( sub {
+         my ( $content ) = @_;
+         log_if_fail "Content", $content;
+
+         assert_json_keys( $content, "first_message_index" );
+
+         assert_json_keys( $content, "forwarded_count" );
+
+         assert_json_keys( $content, "is_verified" );
+
+         assert_json_keys( $content, "session_data" );
+
+         $content->{first_message_index} == 1 or
+            die "Expected first message index to match submitted data";
+
+         $content->{forwarded_count} == 0 or
+            die "Expected forwarded count to match submitted data";
+
+         $content->{is_verified} == JSON::false or
+            die "Expected is_verified to match submitted data";
+
+         $content->{session_data} eq "anotheropaquestring" or
+            die "Expected session data to match submitted data";
+
+         Future->done(1);
+      });
+   };
+
+test "Will not update keys with worse versions",
+   requires => [ $fixture, qw( can_update_e2e_keys ) ],
+
+   proves => [qw( wont_update_e2e_keys )],
+
+   do => sub {
+      my ( $user ) = @_;
+
+      do_request_json_for( $user,
+         method  => "PUT",
+         uri     => "/unstable/room_keys/keys/!abcd/1234",
+         params  => {
+            version => $current_version,
+         },
+         content => {
+            first_message_index => 5,
+            forwarded_count     => 0,
+            is_verified         => JSON::false,
+            session_data        => "yetanotheropaquestring",
+         }
+      )->then( sub {
+         my ( $content ) = @_;
+         log_if_fail "Content", $content;
+
+         do_request_json_for( $user,
+            method  => "GET",
+            uri     => "/unstable/room_keys/keys/!abcd/1234",
+            params  => {
+               version => $current_version,
+            }
+         );
+      })->then( sub {
+         my ( $content ) = @_;
+         log_if_fail "Content", $content;
+
+         assert_json_keys( $content, "first_message_index" );
+
+         assert_json_keys( $content, "forwarded_count" );
+
+         assert_json_keys( $content, "is_verified" );
+
+         assert_json_keys( $content, "session_data" );
+
+         # The data should not be overwritten, so should be the same as what
+         # was set by the previous test.
+         $content->{first_message_index} == 1 or
+            die "Expected first message index to match submitted data";
+
+         $content->{forwarded_count} == 0 or
```

Comment:
we should probably exercise `forwarded_count` here too - perhaps just put another update attempt in here to check that forwarded keys don't trump non-forwarded ones? (iirc that's how the logic went)